grin/src/Transformations/Syntax.hs

```diff
@@ -0,0 +1,164 @@
+  convert (New.ConstTagNode t vars) = ConstTagNode t (map Var vars)
+  convert (New.Unit)  = Unit
+  convert (New.Lit l) = Lit l
+  convert (New.Var v) = Var v
+  convert (New.Undefined t) = Undefined t
+
+instance Convertible New.Exp Exp where
+  convert (New.Program exts defs)  = Program exts (map convert defs)
+  convert (New.Def name args body) = Def name args (convert body)
+  convert e@(New.EBind lhs pat rhs)
+    | (New.VarPat v)      <- pat = EBind (convert lhs) (Var v) (convert rhs)
+    | (New.AsPat  v pat') <- pat -- condition
+    , rhs' <- EBind (SReturn (Var v)) (convert pat') (convert rhs) -- helper
+    = EBind (convert lhs) (Var v) rhs'
+  convert e@(New.ECase scrut alts) = ECase (Var scrut) (map convert alts)
+  convert (New.SApp f vars)        = SApp f $ map Var vars
+  convert (New.SStore var)         = SStore (Var var)
+  convert (New.SFetch ptr)         = SFetchI ptr Nothing
+  convert (New.SUpdate ptr var)    = SUpdate ptr (Var var)
+  convert (New.SReturn val)        = SReturn (convert val)
+  convert (New.SBlock exp)         = SBlock (convert exp)
+  convert (New.Alt cpat exp)       = Alt cpat (convert exp)
+
+convertToNew :: Exp -> New.Exp
+convertToNew = convert . nameEverything
+
+nameEverything :: Exp -> Exp
+nameEverything = nodeArgumentNaming
+               . bindNormalisation
+               . appArgumentNaming
+               . bindNormalisation
+               . fst . bindingPatternSimplification
+               . bindNormalisation
+               . fst . producerNameIntroduction
+               . bindNormalisation
+
+appArgumentNaming :: Exp -> Exp
+appArgumentNaming e = fst . evalNameM e . cata alg $ e where
+  alg :: ExpF (NameM Exp) -> NameM Exp
+  alg e = case e of
+    SAppF f args -> bindFunArgs f args
+    expf -> fmap embed . sequence $ expf
+
+  bindFunArgs :: Name -> [Val] -> NameM Exp
+  bindFunArgs f args = do
+    varArgs <- forM [1..length args] $ \_ ->
+      Var <$> newArgName
+    let g exp (arg, var) = EBind (SReturn arg) var exp
+        boundApp = foldl g (SApp f varArgs) $ zip args varArgs
+    pure $ SBlock boundApp
+
+  newArgName :: NameM Name
+  newArgName = deriveNewName "x"
+
+-- NOTE: we can ssume tha Producer Name Introduction
+-- & Binding Pattern Simplification has already been run
+-- ConstTagNodes can only appear in SReturns
+nodeArgumentNaming :: Exp -> Exp
+nodeArgumentNaming e = fst . evalNameM e . cata alg $ e where
+  alg :: ExpF (NameM Exp) -> NameM Exp
+  alg e = case e of
+    SReturnF (ConstTagNode tag args) -> bindNodeArgs tag args
+    expf -> fmap embed . sequence $ expf
+
+  bindNodeArgs :: Tag -> [Val] -> NameM Exp
+  bindNodeArgs tag args = do
+    varArgs <- forM [1..length args] $ \_ ->
+      Var <$> newArgName
+    let g exp (arg, var) = EBind (SReturn arg) var exp
+        boundApp = foldl g (SReturn $ ConstTagNode tag varArgs) $ zip args varArgs
+    pure $ SBlock boundApp
+
+  newArgName :: NameM Name
+  newArgName = deriveNewName "y"
+
+testToNew :: IO ()
```

Comment:
This should be in its own Hspec test, please move it there.

Reply:
This is just for utility, and it has its own test. I will remove this.